### PR TITLE
Add multiple author support to module!

### PR DESCRIPTION
**Pull request by wcampbell0x2a**
Change author into authors, allowing multiple authors of a module.

See #244

### CODE REVIEW

File: samples/rust/rust_minimal.rs [hunk 1]
```diff
@@ -7,7 +7,7 @@ use kernel::prelude::*;
 module! {
     type: RustMinimal,
     name: "rust_minimal",
-    author: "Rust for Linux Contributors",
+    authors: ["Rust for Linux Contributors", "Example Author"],
```
Comment (wcampbell0x2a) [inline samples/rust/rust_minimal.rs]: I wouldn't need to keep this, was just testing this with `modinfo`.

**Comment by wcampbell0x2a** [comment]
This might also work for #246

File: rust/macros/module.rs [hunk 2]
```diff
@@ -300,6 +300,25 @@ impl ModuleInfo {
...
+    fn parse_authors(it: &mut token_stream::IntoIter) -> Vec<String> {
+        let mut authors: Vec<String> = vec![];
+        let group = expect_group(it);
+        assert_eq!(group.delimiter(), Delimiter::Bracket);
+        let mut stream = group.stream().into_iter();
+        loop {
+            if let Some(s) = try_string(&mut stream) {
+                authors.push(s);
+            }
+            if try_punct(&mut stream).is_none() {
```
Comment (adamrk) [inline rust/macros/module.rs]: Looks like this would accept any punctuation between the authors. We probably just want `,`, right?

File: rust/macros/module.rs [hunk 3]
```diff
@@ -300,6 +300,25 @@ impl ModuleInfo {
...
+                assert_eq!(group.delimiter(), Delimiter::Bracket);
```
Comment (adamrk) [inline rust/macros/module.rs]: Didn't we already check this above?

File: rust/macros/module.rs [hunk 4]
```diff
@@ -300,6 +300,25 @@ impl ModuleInfo {
...
     }
+
+    /// Parse ["First", "Second"] into Some(vec!["First", "Second"])
+    fn parse_authors(it: &mut token_stream::IntoIter) -> Vec<String> {
+        let mut authors: Vec<String> = vec![];
+        let group = expect_group(it);
+        assert_eq!(group.delimiter(), Delimiter::Bracket);
+        let mut stream = group.stream().into_iter();
+        loop {
+            if let Some(s) = try_string(&mut stream) {
```
Comment (adamrk) [inline rust/macros/module.rs]: We can probably `expect_string` here. Otherwise we'd allow extra commas without any authors in between.

Comment (wcampbell0x2a) [reply rust/macros/module.rs]: Fixed, see new commit.

File: rust/macros/module.rs (hunk 3 above)
Comment (wcampbell0x2a) [reply rust/macros/module.rs]: Fixed, see new commit.

File: rust/macros/module.rs (hunk 2 above)
Comment (wcampbell0x2a) [reply rust/macros/module.rs]: Fixed, see new commit.

**Comment by nbdd0121** [comment]
I suppose we could accept both an array or a single string literal?

**Comment by wcampbell0x2a** [comment]
> I suppose we could accept both an array or a single string literal?

@ojeda do you have comments on this?

File: rust/macros/module.rs [hunk 5]
```diff
@@ -300,6 +300,25 @@ impl ModuleInfo {
...
+        let group = expect_group(it);
+        assert_eq!(group.delimiter(), Delimiter::Bracket);
+        let mut stream = group.stream().into_iter();
+        loop {
+            let author = expect_string(&mut stream);
+            authors.push(author);
+            if let Some(punct) = try_punct(&mut stream) {
+                assert_eq!(punct, ',');
+            } else {
+                break;
```
Comment (adamrk) [inline rust/macros/module.rs:317]: I think this will currently parse a stream where `try_punct` fails because the next token isn't a punct. E.g. `["A", "B" "C"]` will successfully parse as `vec!["A", "B"]` because the missing comma before `"C"` breaks us out of the loop. We probably need to check that there isn't anything extra left in the stream too.

**Comment by ojeda** [comment]
> @ojeda do you have comments on this?

It sounds like a good idea, but then again it introduces complexity that is not really needed, so I am ambivalent. Having always the list, even with 1 entry, is simpler to learn and to implement, and makes it obvious there may be several authors for those that may not know about it. `Cargo.toml` does the same with the `authors` key.

**Comment by wcampbell0x2a** [comment]
Updated, let me know if this looks good or needs any other modifications.

**Comment by adamrk** [comment]
Sorry, I missed the recent updates - @ojeda is the process now that these should be submitted to the mailing list?

**Comment by adamrk** [comment]
Also the changes now look good to me.

**Review by 0x4248 — APPROVED** [review]

**Comment by wcampbell0x2a** [comment]
> Sorry, I missed the recent updates - @ojeda is the process now that these should be submitted to the mailing list?

Either way, I can re-submit the patch.

**Comment by ojeda** [comment]
> @ojeda is the process now that these should be submitted to the mailing list?

Yeah, that is right, we are in the process of submitting things piece by piece to upstream. Please see https://rust-for-linux.com/contributing and https://rust-for-linux.com/branches for some details.

File: rust/macros/module.rs [hunk 6]
```diff
@@ -300,6 +300,26 @@ impl ModuleInfo {
 
         info
     }
+
+    /// Parse ["First", "Second"] into Some(vec!["First", "Second"])
```
Comment (ojeda) [inline rust/macros/module.rs:304]:
```suggestion
    /// Parses `["First", "Second"]` into `Some(vec!["First", "Second"])`.
```

i.e. please use Markdown, a period at the end, and the third person.

In addition, this does not add `Some(...)` -- it is the caller that does it, no?

I would reword it like this instead, if you want to give an example:

```suggestion
    /// Parses a list of authors.
    ///
    /// # Examples
    ///
    /// Parsing `["First", "Second"]` returns a `vec!["First", "Second"]`.
```

**Review by ojeda — CHANGES_REQUESTED** [review]
By the way, this should update the example at `rust/macros/lib.rs` too.
